### PR TITLE
[REVIEW] fix/groupby_without_aggregation

### PR DESCRIPTION
**Pull request by aocsa**
List of changes:
- [x]  Fix `gdf_group_by_without_aggregations` for cases when the input keys don't have valids, the output table should not also have valids. 
- [x] More unit tests to cover this cases. 

Note: This PR is needed to be merge first, in order to complete PR #2605, the new sort-based groupby.

### CODE REVIEW

File: cpp/tests/groupby/without_agg/single_column_groupby.cu [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  using Value = V;
+};
+
+using TestingTypes =
+    ::testing::Types<KV<int8_t, int8_t>, KV<int32_t, int32_t>,
+                     KV<int64_t, int64_t>, KV<int32_t, float>,
+                     KV<int32_t, double>, KV<cudf::category, int32_t>,
+                     KV<cudf::date32, int8_t>, KV<cudf::date64, double>>;;
+
+TYPED_TEST_CASE(SingleColumnGroupby, TestingTypes);
```
Comment (jrhemstad) [inline cpp/tests/groupby/without_agg/single_column_groupby.cu]: All of these tests need to be updated to make sense for a groupby w/o agg. They were copied from the other groupby tests but weren't updated to stand alone.

Comment (aocsa) [reply cpp/tests/groupby/without_agg/single_column_groupby.cu]: I updated the unit tests.

File: cpp/src/groupby/groupby_without_aggregation.cu [hunk 2]
```diff
@@ -91,11 +91,22 @@ std::pair<cudf::table, gdf_column> gdf_group_by_without_aggregations(
 
   gdf_size_type nrows = input_table.num_rows();
 
+  // Ask if input table has nulls
+  auto get_key_table = [&]() {
+    std::vector<gdf_column *> key_columns;
```
Comment (harrism) [inline cpp/src/groupby/groupby_without_aggregation.cu]:
```suggestion
    std::vector<gdf_column *> key_columns(num_key_cols);
    std::transform(key_columns.begin(), key_columns.end(), std::begin(key_col_indices), 
                             [&](auto i) { return input_table.get_column(key_col_indices[i]); });
```

File: cpp/tests/groupby/without_agg/single_column_groupby.cu [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+
+TYPED_TEST(SingleColumnGroupby, OneGroupEvenNullKeys) {
+  constexpr int size{10};
+  using Key = typename SingleColumnGroupby<TypeParam>::KeyType;
+  using Value = typename SingleColumnGroupby<TypeParam>::ValueType;
+  
+  Key key{42};
+  // The sum of n odd numbers is n^2
```
Comment (harrism) [inline cpp/tests/groupby/without_agg/single_column_groupby.cu]: I thought this was groupby without aggregations?

Comment (aocsa) [reply cpp/tests/groupby/without_agg/single_column_groupby.cu]: the tests were updated

File: cpp/src/groupby/groupby_without_aggregation.cu (hunk 2 above)
Comment (aocsa) [reply cpp/src/groupby/groupby_without_aggregation.cu]: I changed the lambda function to a simple one raw loop. The tests were updated previously in order to match more accurately the cases which this fix is covering.

Comment (harrism) [reply cpp/src/groupby/groupby_without_aggregation.cu]: Please replace the raw loop with a std::transform as I suggested.  Suggested viewing. https://channel9.msdn.com/Events/GoingNative/2013/Cpp-Seasoning

File: CHANGELOG.md [hunk 4]
```diff
@@ -25,12 +25,12 @@
 - PR #2621 Fixes the floordiv issue of not promoting float type when rhs is 0
 - PR #2611 Types Test: fix static casting from negative int to string
 - PR #2618 IO Readers: Fix datasource memory map failure for multiple reads
+- PR #2628 groupby_without_aggregation: Fix this function for cases when the input table don't have valids, the output table should not also have valids
```
Comment (harrism) [inline CHANGELOG.md]: Briefer: 

```suggestion
- PR #2628 groupby_without_aggregation non-nullable input table produces non-nullable output
```

File: cpp/tests/groupby/without_agg/single_column_groupby.cu [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+      expect_columns_are_equal(*lhs_col, *rhs_col);
+      return true;
+    }
+  };
+
+  void expect_tables_are_equal(cudf::table const &lhs, cudf::table const &rhs) {
+    EXPECT_EQ(lhs.num_columns(), rhs.num_columns());
+    EXPECT_EQ(lhs.num_rows(), rhs.num_rows());
+    EXPECT_TRUE(
+        std::equal(lhs.begin(), lhs.end(), rhs.begin(),
```
Comment (harrism) [inline cpp/tests/groupby/without_agg/single_column_groupby.cu]: Do we not already have an equality comparator for tables?

File: cpp/tests/groupby/without_agg/single_column_groupby.cu [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+  using T = typename SingleColumnGroupby<TypeParam>::Key;
+  bool ignore_null_keys = false;
+
+  this->evaluate_test(
+      column_wrapper<T>({T(0), T(1), T(2), T(3), T(4), T(5), T(6), T(7)},
+                        [](auto index) { return index % 2; }),
+      column_wrapper<T>({T(1), T(3), T(5), T(7), T(0), T(0), T(0), T(0)},
+                        [](auto index) { return index < 4; }), /*  */
+      column_wrapper<gdf_size_type>({0, 1, 2, 3, 4}), ignore_null_keys);
+}
```
Comment (harrism) [inline cpp/tests/groupby/without_agg/single_column_groupby.cu]:
```suggestion
}

```

File: cpp/tests/groupby/without_agg/single_column_groupby.cu (hunk 5 above)
Comment (harrism) [reply cpp/tests/groupby/without_agg/single_column_groupby.cu]: OK, but I don't think it needs to be in this PR.

**Comment by harrism** [comment]
@jrhemstad are you OK to approve?

**Pull request merged**